funsor/distributions.py

```diff
@@ -104,6 +105,27 @@ def eager_log_prob(cls, *params):
         data = cls.dist_class(**params).log_prob(value)
         return Tensor(data, inputs)
 
+    def unscaled_sample(self, sampled_vars, sample_inputs):
+        params = OrderedDict(self.params)
+        value = params.pop("value")
+        assert all(isinstance(v, (Number, Tensor)) for v in params.values())
+        assert isinstance(value, Variable) and value.name in sampled_vars
+        inputs_, tensors = align_tensors(*params.values())
+        raw_dist = self.dist_class(**dict(zip(self._ast_fields[:-1], tensors)))
+        if getattr(raw_dist, "has_rsample", False):
+            raw_sample = raw_dist.rsample(tuple(v.dtype for v in sample_inputs.values()))
+        else:
+            raw_sample = raw_dist.sample(tuple(v.dtype for v in sample_inputs.values()))
+        raw_score_function = raw_dist.score_parts(raw_sample).score_function
```

Comment:
This performs unnecessary work: in case `has_rsample == True` it unnecessarily computes `log_prob`. Could we replace with
```py
if not raw_dist.has_rsample:
    log_prob = raw_dist.log_prob(raw_sample)
    dice_factor = Tensor(log_prob - log_prob.detach(), inputs)
    result += dice_factor
```
If you want to support `Rejector` you could add extra logic like `if type(self).score_parts is Distribution.score_parts)`, but I doubt anyone uses `Rejector`.